### PR TITLE
Adds base functionality for Brave Wallet Dapps support

### PR DESCRIPTION
**Pull request by kylehickinson**
## Summary of Changes

This initial PR is mostly to get a security review going for the base functionality around integrating with an ethereum provider 

Dapps functionality is currently behind a compilation flag `WALLET_DAPPS_ENABLED`, so to test this you must build locally with that Swift flag added in `Base.xcconfig` or manually via project settings

A majority of dapps support is still not fully integrated and ready, so this PR is mostly just for the communications layer between dapps and brave-core

## Submitter Checklist:

- [ ] *Unit Tests* are updated to cover new or changed functionality
- [ ] User-facing strings use `NSLocalizableString()`

## Reviewer Checklist:

- [ ] Issues include necessary QA labels:
  - `QA/(Yes|No)`
  - `bug` / `enhancement`
- [ ] Necessary [security reviews](https://github.com/brave/security/issues/new/choose) have taken place.
- [ ] Adequate unit test coverage exists to prevent regressions.
- [ ] Adequate test plan exists for QA to validate (if applicable).
- [ ] Issue and pull request is assigned to a milestone (should happen at merge time).

### CODE REVIEW

File: Client/Frontend/Browser/BrowserViewController/BrowserViewController+WKNavigationDelegate.swift [hunk 1]
```diff
@@ -468,6 +471,7 @@ extension BrowserViewController: WKNavigationDelegate {
           isPrivate: PrivateBrowsingManager.shared.isPrivateBrowsing
         )
       }
+            tab.updateEthereumProperties()
```
Comment (StephenHeaps) [inline Client/Frontend/Browser/BrowserViewController/BrowserViewController+WKNavigationDelegate.swift]: Indent in this function is off

File: Client/Frontend/Browser/BrowserViewController/BrowserViewController+ToolbarDelegate.swift [hunk 2]
```diff
@@ -275,6 +275,10 @@ extension BrowserViewController: TopToolbarDelegate {
     scanQRCode()
   }
 
+    func topToolbarDidTapWalletButton(_ urlBar: TopToolbarView) {
+        presentWalletPanel()
```
Comment (StephenHeaps) [inline Client/Frontend/Browser/BrowserViewController/BrowserViewController+ToolbarDelegate.swift]: Indent

File: Client/Frontend/Browser/Toolbars/UrlBar/TopToolbarView.swift [hunk 3]
```diff
@@ -48,6 +48,7 @@ protocol TopToolbarDelegate: AnyObject {
   func topToolbarDidPressStop(_ urlBar: TopToolbarView)
   func topToolbarDidPressReload(_ urlBar: TopToolbarView)
   func topToolbarDidPressQrCodeButton(_ urlBar: TopToolbarView)
+    func topToolbarDidTapWalletButton(_ urlBar: TopToolbarView)
```
Comment (StephenHeaps) [inline Client/Frontend/Browser/Toolbars/UrlBar/TopToolbarView.swift]: Indent

File: Client/Frontend/Browser/BrowserViewController/BrowserViewController+Wallet.swift [hunk 4]
```diff
@@ -19,3 +73,175 @@ extension BrowserViewController: BraveWalletDelegate {
...
+
+extension Tab: BraveWalletEventsListener {
+  func chainChangedEvent(_ chainId: String) {
+    Task { @MainActor in
+      guard let provider = walletProvider,
+            case let currentChainId = await provider.chainId(),
+            chainId != currentChainId else { return }
+      webView?.evaluateSafeJavaScript(
+        functionName: "window.ethereum.emit",
+        args: ["chainChanged", chainId],
```
Comment (StephenHeaps) [inline Client/Frontend/Browser/BrowserViewController/BrowserViewController+Wallet.swift]: I think we meant to store `chainChanged` and `accountsChanged` as constants somewhere

File: BraveUI/Popover/PopoverController.swift [hunk 5]
```diff
@@ -417,7 +417,7 @@ extension PopoverController {
 
     containerView.transform = CGAffineTransform(translationX: translationDelta.x, y: translationDelta.y)
       .scaledBy(x: scale, y: scale)
-      .rotated(by: rotation)
+//            .rotated(by: rotation)
```
Comment (StephenHeaps) [inline BraveUI/Popover/PopoverController.swift:420]: I think we could add a flag to enable/disable rotation for the `PopoverController`

File: Client/Frontend/Browser/BrowserViewController.swift [hunk 6]
```diff
@@ -92,8 +92,8 @@ class BrowserViewController: UIViewController, BrowserViewControllerDelegate {
   var updateDisplayedPopoverProperties: (() -> Void)?
 
   let profile: Profile
-  let tabManager: TabManager
   let braveCore: BraveCoreMain
+    let tabManager: TabManager
```
Comment (StephenHeaps) [inline Client/Frontend/Browser/BrowserViewController.swift]: Indent

**Review by StephenHeaps — COMMENTED** [review]
Mostly just a few spots to fix indent 🙂

File: Client/Frontend/Browser/BrowserViewController/BrowserViewController+Wallet.swift [hunk 7]
```diff
@@ -19,3 +73,175 @@ extension BrowserViewController: BraveWalletDelegate {
...
+        completion: nil
+      )
+      let networkVersion = valueOrUndefined(Int(chainId.removingHexPrefix, radix: 16))
+      webView.evaluateSafeJavaScript(
+        functionName: "window.ethereum.networkVersion = \(networkVersion)",
+        contentWorld: .page,
+        asFunction: false,
+        completion: nil
+      )
+      let selectedAccount = valueOrUndefined(await allowedAccounts(false).0)
```
Comment (nuo-xu) [inline Client/Frontend/Browser/BrowserViewController/BrowserViewController+Wallet.swift]: i'm little confused about the `allowedAccounts` which are from database and the `accounts` from `accountsChangedEvent`. Shouldnt we update eth properties with the `accounts` from `accountsChangedEvent`?

Comment (kylehickinson) [reply Client/Frontend/Browser/BrowserViewController/BrowserViewController+Wallet.swift]: Discussed in Slack, following desktops flow, which is just to fetch allowed accounts instead of taking whats in the events listener

File: BraveUI/Popover/PopoverController.swift (hunk 5 above)
Comment (kylehickinson) [reply BraveUI/Popover/PopoverController.swift:420]: we've had a bunch of problems with rotation here (due to AutoLayout + safe area problems), I think it may be better to just fully remove it. Will bring it up with rest of iOS team

File: BraveWallet/Panels/Add Suggested Token/AddSuggestedTokenView.swift [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+        }
+        .frame(maxWidth: .infinity)
+        .padding(.top, 64)
+        .padding([.leading, .trailing, .bottom])
+      }
+      .background(Color(.braveGroupedBackground).ignoresSafeArea())
+      .toolbar {
+        ToolbarItemGroup(placement: .cancellationAction) {
+          Button {
+
```
Comment (iccub) [inline BraveWallet/Panels/Add Suggested Token/AddSuggestedTokenView.swift:45]: would it make sense to mark all those TBA button actions to not forget them when you submit next pull request?

File: Client/Frontend/Browser/UserScriptManager.swift [hunk 9]
```diff
@@ -407,6 +434,15 @@ class UserScriptManager {
       if let domainUserScript = domainUserScript, let script = domainUserScript.script {
         $0.addUserScript(script)
       }
+
+      #if WALLET_DAPPS_ENABLED
+      if let script = walletProviderScript, !PrivateBrowsingManager.shared.isPrivateBrowsing {
```
Comment (iccub) [inline Client/Frontend/Browser/UserScriptManager.swift]: perhaps better to check for `tab?.isPrivate` instead of the global state?

File: Client/Wallet/EthereumProviderHelper.swift [hunk 10]
```diff
@@ -0,0 +1,170 @@
...
+  func userContentController(
+    _ userContentController: WKUserContentController,
+    didReceiveScriptMessage message: WKScriptMessage,
+    replyHandler: @escaping (Any?, String?) -> Void
+  ) {
+    guard let tab = tab,
+          let provider = tab.walletProvider,
+          !message.frameInfo.securityOrigin.host.isEmpty, // Fail if there is no last committed URL yet
+          message.frameInfo.isMainFrame, // Fail the request came from 3p origin
+          JSONSerialization.isValidJSONObject(message.body),
```
Comment (iccub) [inline Client/Wallet/EthereumProviderHelper.swift:74]: we check for private mode in few other places, in case we want to add additional check here we can do so by looking at `tab.isPrivate` property

File: package.json [hunk 11]
```diff
@@ -14,7 +14,8 @@
     "brave-core-ios": "https://github.com/brave/brave-browser/releases/download/v1.38.73/brave-core-ios-1.38.73.tgz",
     "page-metadata-parser": "^1.1.3",
     "readability": "github:mozilla/readability#b9f47bcc8d3c223cabe2dec6a42eeb3bd778d85c",
-    "webpack-cli": "^4.8.0"
+    "webpack-cli": "^4.8.0",
+    "events": "^3.3.0"
```
Comment (iccub) [inline package.json]: Maybe worth mentioning in the security review that we add this dep?
from package-lock looks like we use it already though

File: Client/Frontend/Browser/BrowserViewController.swift [hunk 12]
```diff
@@ -1980,10 +1987,13 @@ extension BrowserViewController: TabDelegate {
     readerMode.delegate = self
     tab.addContentScript(readerMode, name: ReaderMode.name(), contentWorld: .defaultClient)
 
-    // only add the logins helper if the tab is not a private browsing tab
+        // only add the logins helper and wallet provider if the tab is not a private browsing tab
     if !tab.isPrivate {
       let logins = LoginsHelper(tab: tab, profile: profile, passwordAPI: braveCore.passwordAPI)
       tab.addContentScript(logins, name: LoginsHelper.name(), contentWorld: .defaultClient)
+      #if WALLET_DAPPS_ENABLED
```
Comment (iccub) [inline Client/Frontend/Browser/BrowserViewController.swift:2049]: We usually put such flags as a negation and versioned them,
does not matter for me personally as this flag will be dropped really soon

**Review by iccub — APPROVED** [review]
looks good to me

File: package.json (hunk 11 above)
Comment (kylehickinson) [reply package.json]: Oh, this can actually be removed! I forgot I added it prior to using 1.38.x and being able to grab the ethereum provider JS from brave-core

File: BraveWallet/Panels/Add Suggested Token/AddSuggestedTokenView.swift (hunk 8 above)
Comment (kylehickinson) [reply BraveWallet/Panels/Add Suggested Token/AddSuggestedTokenView.swift:45]: They should be covered by the open issues surrounding dapps (such as #5104)

Overall would have been better if I split this PR up more

File: BraveWallet/Crypto/CryptoView.swift [hunk 13]
```diff
@@ -61,11 +69,37 @@ public struct CryptoView: View {
...
+              )
+            case .webpageRequests:
+              WebpageRequestContainerView(
+                keyringStore: keyringStore,
+                cryptoStore: store,
+                toolbarDismissContent: dismissButtonToolbarContents
+              )
+            case .requestEthererumPermissions(let request):
+              NewSiteConnectionView(
+                origin: request.requestingOrigin,
```
Comment (thypon) [inline BraveWallet/Crypto/CryptoView.swift:88]: Is the TLD+1/2 always visible and marked in bold (maybe even yellow/gold) when the origin is super long?

File: BraveWallet/Panels/Add Suggested Network/AddSuggestedNetworkView.swift [hunk 14]
```diff
@@ -0,0 +1,136 @@
...
+  @ScaledMetric private var blockieSize = 24
+  @ScaledMetric private var faviconSize = 48
+  
+  @Environment(\.openWalletURLAction) private var openWalletURL
+  
+  private var headerView: some View {
+    VStack {
+      HStack(spacing: 8) {
+        Spacer()
+        Text(keyringStore.selectedAccount.address.truncatedAddress)
```
Comment (thypon) [inline BraveWallet/Panels/Add Suggested Network/AddSuggestedNetworkView.swift:26]: Is the truncated address here the `origin` as well?

Comment (thypon) [inline BraveWallet/Panels/Add Suggested Network/AddSuggestedNetworkView.swift:26]: Is the truncated address here the `origin` as well?

File: BraveWallet/Panels/Connect/NewSiteConnectionView.swift [hunk 15]
```diff
@@ -0,0 +1,209 @@
...
+  @State private var selectedAccounts: Set<BraveWallet.AccountInfo.ID> = []
+  @State private var isConfirmationViewVisible: Bool = false
+  
+  private var headerView: some View {
+    VStack(spacing: 8) {
+      Image(systemName: "globe")
+        .frame(width: faviconSize, height: faviconSize)
+        .background(Color(.braveDisabled))
+        .clipShape(RoundedRectangle(cornerRadius: 4, style: .continuous))
+      Text(verbatim: origin.absoluteString)
```
Comment (thypon) [inline BraveWallet/Panels/Connect/NewSiteConnectionView.swift:43]: Is it possible to bold and change the color for TLD+1/2?

Comment (thypon) [inline BraveWallet/Panels/Connect/NewSiteConnectionView.swift:43]: Is it possible to bold and change the color for TLD+1/2?

File: Client/Wallet/EthereumProviderHelper.swift [hunk 16]
```diff
@@ -0,0 +1,171 @@
...
+    else {
+      log.error("Failed to handle ethereum provider communication")
+      return
+    }
+    
+    // The web page has communicated with `window.ethereum`, so we should show the wallet icon
+    tab.isWalletIconVisible = true
+    
+    let securityOrigin = message.frameInfo.securityOrigin
+    let origin = "\(securityOrigin.protocol)://\(securityOrigin.host):\(securityOrigin.port)"
```
Comment (thypon) [inline Client/Wallet/EthereumProviderHelper.swift:87]: Are we checking the origins after the message is sent, where?

File: BraveWallet/Crypto/CryptoView.swift (hunk 13 above)
Comment (kylehickinson) [reply BraveWallet/Crypto/CryptoView.swift:88]: It will currently wrap to as many lines as needed if its long

File: BraveWallet/Panels/Add Suggested Network/AddSuggestedNetworkView.swift (hunk 14 above)
Comment (kylehickinson) [reply BraveWallet/Panels/Add Suggested Network/AddSuggestedNetworkView.swift:26]: No, the origin in this PR is hard-coded and not setup, but it will be added here: https://github.com/brave/brave-ios/pull/5164/files#diff-2e77a33043f377f9a58af58bcbb41bbc53d06985612b101c42062a1abe146bc0R90

It will be the same as all other origins we display: multilined/wrapped to display entire string

Comment (kylehickinson) [reply BraveWallet/Panels/Add Suggested Network/AddSuggestedNetworkView.swift:26]: Same comment as above?

Comment (thypon) [reply BraveWallet/Panels/Add Suggested Network/AddSuggestedNetworkView.swift:26]: How many lines will it contain? Does it wrap and scroll if too many are contained?

Comment (kylehickinson) [reply BraveWallet/Panels/Add Suggested Network/AddSuggestedNetworkView.swift:26]: As many as it needs to display the entire address, all UI is contained inside scrollable containers by default

File: BraveWallet/Panels/Connect/NewSiteConnectionView.swift (hunk 15 above)
Comment (kylehickinson) [reply BraveWallet/Panels/Connect/NewSiteConnectionView.swift:43]: Technically possible yeah but possibly out of scope for this PR and possibly something to add to https://github.com/brave/brave-ios/pull/5164 and others

File: Client/Wallet/EthereumProviderHelper.swift (hunk 16 above)
Comment (kylehickinson) [reply Client/Wallet/EthereumProviderHelper.swift:87]: At this point the origin is sent directly into brave-core and handled via the provider implementation https://github.com/brave/brave-core/blob/31c31e76c8899ca71b0314c087ed134da065bd48/components/brave_wallet/browser/brave_wallet_provider_impl.cc#L831

After that it may or may not call `ShowPanel` and we simply fetch pending requests in a particular order, but it isn't tied to a specific tab

Comment (thypon) [reply Client/Wallet/EthereumProviderHelper.swift:87]: What happens if the communication is started from an iframe? Is it blocked or does it start from the tab page?

File: BraveWallet/Panels/Signature Request/SignatureRequestView.swift [hunk 17]
```diff
@@ -0,0 +1,151 @@
...
+      }
+    } else {
+      HStack {
+        buttons
+      }
+    }
+  }
+  
+  @ViewBuilder private var buttons: some View {
+//    Button(action: {
```
Comment (kdenhartog) [inline BraveWallet/Panels/Signature Request/SignatureRequestView.swift:129]: Is this just temporarily being commented out or can it be removed now?

File: BraveWallet/WalletHostingViewController.swift [hunk 18]
```diff
@@ -40,53 +57,57 @@ public class WalletHostingViewController: UIHostingController<CryptoView> {
...
-
+  
   public override var shouldAutorotate: Bool {
     true
   }
 }
 
-private class WalletInteractionGestureRecognizer: UIGestureRecognizer {
+class WalletInteractionGestureRecognizer: UIGestureRecognizer {
   private let store: KeyringStore
```
Comment (kdenhartog) [inline BraveWallet/WalletHostingViewController.swift:111]: nitpick: I'm just seeing this now even though it's not part of this change can we rename this variable to keyringStorage since "store" seems to become an overloaded term very quickly. It seems that would be consistent with how we've done this in other files as well.

File: BraveWallet/WalletProviderPermissionRequestsManager.swift [hunk 19]
```diff
@@ -0,0 +1,77 @@
...
+  public enum Response {
+    /// The user rejected the prompt by dismissing the screen
+    case rejected
+    /// The user granted access to a set of accounts
+    case granted(accounts: [String])
+  }
+  /// The origin that requested this permission
+  let requestingOrigin: URL
+  /// The type of request
+  let coinType: BraveWallet.CoinType
```
Comment (kdenhartog) [inline BraveWallet/WalletProviderPermissionRequestsManager.swift:22]: Is this a required parameter for every request under our current usage? For example, I saw that we're supporting signMessage request which I believe wouldn't require any coinType. Should this parameter be defined as an optional usage in this struct?

File: BraveWallet/WalletStrings.swift [hunk 20]
```diff
@@ -1023,12 +1023,12 @@ extension Strings {
...
     )
-    public static let nextTransaction = NSLocalizedString(
-      "wallet.nextTransaction",
+    public static let next = NSLocalizedString(
+      "wallet.next",
       tableName: "BraveWallet",
       bundle: .braveWallet,
       value: "Next",
-      comment: "A button title next indicating the user to go to the next transaction. Will sit next to a label such as \"1 of 4\" where tapping next would move them to the second transaction"
+      comment: "1. A button title next indicating the user to go to the next transaction. Will sit next to a label such as \"1 of 4\" where tapping next would move them to the second transaction. 2. Title of the button for users to click to the next step during the process of dapp permission requests."
```
Comment (kdenhartog) [inline BraveWallet/WalletStrings.swift:1031]: Do we have a maximum line length setup in the linter?

File: Client/Frontend/Browser/BrowserViewController/BrowserViewController+Wallet.swift [hunk 21]
```diff
@@ -6,9 +6,64 @@
...
+import SwiftUI
+import BraveUI
+import Data
+import BraveShared
+
+private let log = Logger.browserLogger
+
+extension WalletStore {
+  /// Creates a WalletStore based on whether or not the user is in Private Mode
+  static func from(privateMode: Bool) -> WalletStore? {
```
Comment (kdenhartog) [inline Client/Frontend/Browser/BrowserViewController/BrowserViewController+Wallet.swift:20]: Just for my knowledge is privateMode something like using incognito mode or something else?

File: Client/Frontend/Browser/Tab.swift [hunk 22]
```diff
@@ -66,6 +66,9 @@ class Tab: NSObject {
 
   var secureContentState: TabSecureContentState = .unknown
 
+  var walletProvider: BraveWalletBraveWalletProvider?
```
Comment (kdenhartog) [inline Client/Frontend/Browser/Tab.swift:69]:
```suggestion
var walletProvider: BraveWalletProvider?
```

I think this wasn't meant to be listed twice

File: Client/Frontend/Browser/TabManager.swift [hunk 23]
```diff
@@ -81,6 +82,7 @@ class TabManager: NSObject {
   var selectedIndex: Int { return _selectedIndex }
   var tempTabs: [Tab]?
   private weak var rewards: BraveRewards?
+  var makeWalletProvider: ((Tab) -> (BraveWalletBraveWalletProvider, js: String)?)?
```
Comment (kdenhartog) [inline Client/Frontend/Browser/TabManager.swift:85]: Hmm this being used here a second time makes me think it may have been intentional. Is this a pattern that was deliberately chosen?

File: Client/Frontend/UserContent/UserScripts/WalletEthereumProvider.js [hunk 24]
```diff
@@ -0,0 +1,89 @@
...
+    request: function (args) /* -> Promise<unknown> */  {
+      return post('request', args)
+    },
+    isConnected: function() /* -> bool */ {
+      return true;
+    },
+    enable: function() /* -> void */ {
+      return post('enable', {})
+    },
+    // ethereum.sendAsync(payload: JsonRpcRequest, callback: JsonRpcCallback): void;
```
Comment (kdenhartog) [inline Client/Frontend/UserContent/UserScripts/WalletEthereumProvider.js:40]: Does this project use JS Docs at all? Thinking that might help for defining these API types.

File: Client/Frontend/UserContent/UserScripts/WalletEthereumProvider.js [hunk 25]
```diff
@@ -0,0 +1,89 @@
...
+      try {
+        reject(JSON.parse(errorJSON))
+      } catch(e) {
+        reject(errorJSON)
+      }
+    })
+  })
+}
+
+Object.defineProperty(window, 'ethereum', {
```
Comment (kdenhartog) [inline Client/Frontend/UserContent/UserScripts/WalletEthereumProvider.js:25]: Is it possible for us to reuse the object definition in brave-core here: https://github.com/brave/brave-core/blob/8385231a90a9d9cb20ac98b0d96cf58bcffec20e/components/brave_wallet/resources/ethereum_provider.js

File: Client/Frontend/UserContent/UserScripts/WalletEthereumProvider.js [hunk 26]
```diff
@@ -0,0 +1,89 @@
...
+  configurable: true,
```
Comment (kdenhartog) [inline Client/Frontend/UserContent/UserScripts/WalletEthereumProvider.js:26]: If we can't reuse the brave-core definition of this property can we match this object to what's being defined there in order to keep things as closely aligned as possible and make this a bit more maintainable. For example the `writable` and `enumerable` parameters are being set to true rather than the default of false. Ultimately we should align with the goal of making it not possible for a web page to overwrite it like we did with the [solana provider PR](https://github.com/brave/brave-core/pull/12554#discussion_r833722481).

Given this is making RPC calls but brave-core isn't I'm not sure if it's possible to do this.

**Comment by kdenhartog** [comment]
Spotted a few things I had questions about to gain a better understanding of what's going on. I've got a basic build done and want to spend a bit more time trying things out with the `WALLET_DAPPS_ENABLED` enabled to see how it behaves.

Nothing obviously wrong sticks out to me at this point from what I'm seeing, but I want to spend a bit more time trying to overwrite the object once I get it built with the flag enabled

**Review by kdenhartog — COMMENTED** [review]
Spotted a few things I had questions about to gain a better understanding of what's going on. I've got a basic build done and want to spend a bit more time trying things out with the `WALLET_DAPPS_ENABLED` enabled to see how it behaves before approving.

Nothing obviously wrong sticks out to me at this point from what I'm seeing, but I want to spend a bit more time trying to overwrite the object once I get it built with the flag enabled

File: BraveWallet/Panels/Signature Request/SignatureRequestView.swift (hunk 17 above)
Comment (kylehickinson) [reply BraveWallet/Panels/Signature Request/SignatureRequestView.swift:129]: This is already implemented in a different PR: https://github.com/brave/brave-ios/pull/5207/files

File: BraveWallet/WalletProviderPermissionRequestsManager.swift (hunk 19 above)
Comment (kylehickinson) [reply BraveWallet/WalletProviderPermissionRequestsManager.swift:22]: This is only for connection requests, since we can't use the Chromium permissions system. Coin type is added here since we will eventually add Solana and others

File: BraveWallet/WalletStrings.swift (hunk 20 above)
Comment (kylehickinson) [reply BraveWallet/WalletStrings.swift:1031]: We do, but string files are ignored

File: Client/Frontend/Browser/BrowserViewController/BrowserViewController+Wallet.swift (hunk 21 above)
Comment (kylehickinson) [reply Client/Frontend/Browser/BrowserViewController/BrowserViewController+Wallet.swift:20]: Yep, It is the name for incognito mode in iOS

File: Client/Frontend/Browser/Tab.swift (hunk 22 above)
Comment (kylehickinson) [reply Client/Frontend/Browser/Tab.swift:69]: It is correct, and is this way because Obj-C doesn't support namespacing and the [interface defined in the mojom file in brave-core includes "BraveWallet"](https://github.com/brave/brave-core/blob/8532627de764fdb25045255eb856c9c71f8e75c8/components/brave_wallet/common/brave_wallet.mojom#L125). In brave-core the generated C++ interface for this is `brave_wallet::mojom::BraveWalletProvider`. In Obj-C we namespace via name prefix BraveWallet (the mojom module name) to each generated type.  We try to rename this in Swift to `BraveWallet.{x}` but unfortunately protocols cannot be nested inside other types in Swift so we are forced to use the full prefixed name here

File: Client/Frontend/UserContent/UserScripts/WalletEthereumProvider.js (hunk 24 above)
Comment (kylehickinson) [reply Client/Frontend/UserContent/UserScripts/WalletEthereumProvider.js:40]: No JS docs usage, these comments were just to help understand the `window.ethereum` methods we were manually defining

File: Client/Frontend/UserContent/UserScripts/WalletEthereumProvider.js (hunk 25 above)
Comment (kylehickinson) [reply Client/Frontend/UserContent/UserScripts/WalletEthereumProvider.js:25]: That is already used from brave-core & injected, but it serves a different purpose. This JS file is the iOS equivalent of https://github.com/brave/brave-core/blob/master/components/brave_wallet/renderer/js_ethereum_provider.cc

File: Client/Wallet/EthereumProviderHelper.swift (hunk 16 above)
Comment (kylehickinson) [reply Client/Wallet/EthereumProviderHelper.swift:87]: The communication JS & ethereum provider from brave-core  is only injected into main frame, so `window.ethereum` will not exist on iframes.

File: Client/Frontend/UserContent/UserScripts/WalletEthereumProvider.js (hunk 26 above)
Comment (kylehickinson) [reply Client/Frontend/UserContent/UserScripts/WalletEthereumProvider.js:26]: One of the main issues is that the brave-core ethereum provider itself overwrites `window.ethereum` (to be a Proxy) and adds additional properties to it.  Seen here: https://github.com/brave/brave-core/blob/8532627de764fdb25045255eb856c9c71f8e75c8/components/brave_wallet/resources/ethereum_provider.js#L39

Additionally the that redefinition sets `enumerable`, `configurable` and `writable` to true

File: Client/Frontend/UserContent/UserScripts/WalletEthereumProvider.js (hunk 24 above)
Comment (kdenhartog) [reply Client/Frontend/UserContent/UserScripts/WalletEthereumProvider.js:40]: Cool if JS docs aren't normally in use I don't think it makes sense to start now. Happy to leave the comments as is because I did find them helpful for figuring out the expected typing.

File: BraveWallet/WalletProviderPermissionRequestsManager.swift (hunk 19 above)
Comment (kdenhartog) [reply BraveWallet/WalletProviderPermissionRequestsManager.swift:22]: Ok, it doesn't seem to present any security concerns for me at this point so no further action would be needed on this one.

File: BraveWallet/Panels/Signature Request/SignatureRequestView.swift (hunk 17 above)
Comment (kdenhartog) [reply BraveWallet/Panels/Signature Request/SignatureRequestView.swift:129]: Sweet thanks that makes sense to me since you split this one up to make it a bit easier to review

File: BraveWallet/WalletStrings.swift (hunk 20 above)
Comment (kdenhartog) [reply BraveWallet/WalletStrings.swift:1031]: Thanks for the info no further action needed here either

File: Client/Frontend/Browser/BrowserViewController/BrowserViewController+Wallet.swift (hunk 21 above)
Comment (kdenhartog) [reply Client/Frontend/Browser/BrowserViewController/BrowserViewController+Wallet.swift:20]: 👍🏻 thanks for confirming that for me

File: Client/Frontend/Browser/Tab.swift (hunk 22 above)
Comment (kdenhartog) [reply Client/Frontend/Browser/Tab.swift:69]: Ahh very interesting I'll take a deeper look into this for my own context, but I don't see how this could raise any additional concerns so I'd think this wouldn't need any other actions to be taken

File: Client/Frontend/Browser/TabManager.swift (hunk 23 above)
Comment (kdenhartog) [reply Client/Frontend/Browser/TabManager.swift:85]: Same as above, no additional action needed here

File: package.json (hunk 11 above)
Comment (thypon) [reply package.json]: @diracdeltas shouldn't we use precise versions in our `package.json` to avoid package takeover issues?

Comment (kylehickinson) [reply package.json]: The events package here is actually no longer included so there's no change here anymore.

We can always lock in precise versions of everything in our package.json though in a different PR if needed :)

File: Client/Frontend/UserContent/UserScripts/WalletEthereumProvider.js (hunk 25 above)
Comment (kdenhartog) [reply Client/Frontend/UserContent/UserScripts/WalletEthereumProvider.js:25]: Ok thanks for clarifying - I don't think this would be an issue then.

File: Client/Frontend/UserContent/UserScripts/WalletEthereumProvider.js (hunk 26 above)
Comment (kdenhartog) [reply Client/Frontend/UserContent/UserScripts/WalletEthereumProvider.js:26]: @diracdeltas what's your take on this one? Looking at the comment you left on that other PR in the past we could leave this as is and assume it's untrusted, but it seems a bit odd to be able to trust this on desktop/android but not iOS.

Comment (diracdeltas) [reply Client/Frontend/UserContent/UserScripts/WalletEthereumProvider.js:26]: just to make sure i understand, are you referring to this comment? https://github.com/brave/brave-core/pull/12554#discussion_r834805673

to clarify, my concern there was around trusted code in brave-core which called `window.ethereum` or `window.solana` so if a page could redefine it or its properties, that could cause security problems. i think at least for solana there is just one instance of this and it's guarded by another check, as mentioned in https://github.com/brave/brave-core/pull/12554#discussion_r835688146. however there is currently no process/check in place (other than manual code review) which prevents people from adding potentially unsafe code in the future, so it's better if we either make sure these objects cannot be redefined or clearly mark them as untrusted.

Comment (kdenhartog) [reply Client/Frontend/UserContent/UserScripts/WalletEthereumProvider.js:26]: > just to make sure i understand, are you referring to this comment? brave/brave-core#12554 (comment)

Yeah that was the comment I was intending to reference.

> to clarify, my concern there was around trusted code in brave-core which called window.ethereum or window.solana so if a page could redefine it or its properties, that could cause security problems.

Ahh ok that's making more sense to me now why that's a different case then this one. When I originally raised this question I was thinking about this code incorrectly. Based on what I understand now if someone were to override this they'd essentially just be breaking the link here for the JS code to call down to swift code which ultimately calls the brave core code. (and that calls some more JS code to complete the circle of languages 😆 )

In that case I think the solution to preventing this would be what's defined [here](https://github.com/brave/brave-core/pull/12554#discussion_r834841397), but realistically speaking overwriting this would either cause it to break the communication or allow someone to redirect to a different provider. I suppose someone could have some fun if they overwrote the isConnected function to break dApp functionality.

Comment (diracdeltas) [reply Client/Frontend/UserContent/UserScripts/WalletEthereumProvider.js:26]: >  I suppose someone could have some fun if they overwrote the isConnected function to break dApp functionality.

wouldn't they just break it for their own site though? it would be a bad problem if a site could override `isConnected` for other origins.

Comment (kylehickinson) [reply Client/Frontend/UserContent/UserScripts/WalletEthereumProvider.js:26]: They would only break their own site, correct.

Comment (kdenhartog) [reply Client/Frontend/UserContent/UserScripts/WalletEthereumProvider.js:26]: > > I suppose someone could have some fun if they overwrote the isConnected function to break dApp functionality.
> 
> wouldn't they just break it for their own site though? it would be a bad problem if a site could override `isConnected` for other origins.

Yeah, that's my understanding as well now is this is site isolated so shouldn't cause any problems in that way. I think this one is a non-issue now.

**Review by kdenhartog — APPROVED** [review]
LGTM - approving on behalf of the security team after @thypon said it looks good to him in the security review issue

File: Client/Frontend/Browser/BrowserViewController.swift [hunk 27]
```diff
@@ -2033,10 +2042,13 @@ extension BrowserViewController: TabDelegate {
     readerMode.delegate = self
     tab.addContentScript(readerMode, name: ReaderMode.name(), contentWorld: .defaultClient)
 
-    // only add the logins helper if the tab is not a private browsing tab
+        // only add the logins helper and wallet provider if the tab is not a private browsing tab
```
Comment (StephenHeaps) [inline Client/Frontend/Browser/BrowserViewController.swift:2045]: nitpick: indent

File: Client/Frontend/Browser/BrowserViewController.swift [hunk 28]
```diff
@@ -2372,6 +2384,7 @@ extension BrowserViewController: TabManagerDelegate {
     }
 
     updateInContentHomePanel(selected?.url as URL?)
+        updateURLBarWalletButton()
```
Comment (StephenHeaps) [inline Client/Frontend/Browser/BrowserViewController.swift:2387]: nitpick: indent

File: Client/Frontend/Browser/BrowserViewController/BrowserViewController+WKNavigationDelegate.swift [hunk 29]
```diff
@@ -63,6 +63,9 @@ extension BrowserViewController: WKNavigationDelegate {
       return
     }
 
+        tabManager.selectedTab?.isWalletIconVisible = false
+        updateURLBarWalletButton()
```
Comment (StephenHeaps) [inline Client/Frontend/Browser/BrowserViewController/BrowserViewController+WKNavigationDelegate.swift:67]: nitpick: indent

File: BraveUI/Popover/PopoverController.swift (hunk 5 above)
Comment (StephenHeaps) [reply BraveUI/Popover/PopoverController.swift:420]: I think we're can remove these rotations from `PopoverController` now?

**Review by StephenHeaps — APPROVED** [review]
🙌
